doc/doxygen/src/driver-guide.md

```diff
@@ -0,0 +1,381 @@
+'intialize->ready' paradigm, meaning, that device drivers should be ready to use
+and in operation right after they have been initialized. On top, devices should
+work with physical values where ever possible. So e.g. sensors should return
+already converted values in some physical unit instead of RAW data, so that
+users can work directly with data from different devices directly without having
+to care for device specific conversion.
+
+Additionally towards ease of use, all device drivers in RIOT should provide a
+similar 'look and feel'. They should behave similar concerning things like their
+state after initialization, like their used data representation and so on.
+
+Secondly, all device drivers should be optimized for minimal RAM/ROM usage, as
+RIOT targets (very) constrained devices. This implies, that instead of exposing
+all thinkable functionality, the drivers should focus exporting and implementing
+a device's core functionality, thus covering ~95% of the use cases. Furthermore
+great care should be put into ...(?)
+
+Third, it should always be possible, to handle more than a single device of one
+kind. Drivers and their interfaces are thus designed to keep their state
+information in a parameterized location instead of driver defined global
+variables.
+
+Fourth, RIOT defines high-level interfaces for certain groups of devices (i.e.
+netdev for network devices, SAUL for sensors and actuators), which enables users
+to work with a wide variety of devices without having to know anything about the
+actual device that is mapped.
+
+Fifth, during initialization we make sure that we can communicate with a device.
+Other functions should check the dev pointer is not void, and should also handle
+error return values from the lower layer peripheral driver implementations,
+where there are some.
+
+Sixth, device drivers SHOULD be implemented independent of any CPU/board code.
+To achieve this, the driver implementations should be strictly be based on
+platform independent interfaces as the peripheral drivers, xtimer, etc.
+
+# General                                               {#driver-guide-general}
+
+## Documentation                                            {#driver-guide-doc}
+
+Document what your driver does! Most device come with a very large number of
+features, while the corresponding device driver only supports a sub-set of them.
+This should be clearly stated in the device driver's documentation, so that
+anyone wanting to use the driver can find out the supported features without
+having to scan through the code.
+
+## Device descriptor and parameter configuration          {#driver-guide-types}
+
+Each device MUST supply a data structure, holding the devices state and
+configuration, using the naming scheme of `DEVNAME_t` (e.g. `dht_t`, or
+`at86rf2xx_t`). In context of RIOT, we call this data structure the device
+descriptor.
+
+This device descriptor MUST contain all the state data of a device. By this, we
```

Comment:
```suggestion
This device descriptor MUST contain all the state data of a device. This way, we
```

Reply:
I also agree as this sounds better. But again I think this might be better to do in a follow up.